README
 -*- mode: text; coding: utf-8 -*-
2013-06-26  Matthias Kievernagel  mkiever/at/tlink/dot/de
2013-09-25  Matthias Kievernagel  mkiever/at/tlink/dot/de



README for sdl unicode font view
--------------------------------

These are the sources for unicodeview. A unicode font viewer
for the SDL platform.

This program is Copyright 2013 Matthias Kievernagel.
It is distributed under the GPL version 2 license.
For details see the file COPYING.

The distribution contains modified versions of the files
  SDL_ttf.h and SDL_ttf.c
from the SDL_ttf package (version 2.0.11)
which are Copyright (C) 1997-2012 Sam Lantinga <slouken@libsdl.org>
See COPYING_SDL_ttf for details about their license.

This is version 0.03. It is a very rough tool to look
at the glyphs in a font using SDL. The gridded display of 16x16
was chosen to accomodate unifont (http://czyborra.com/unifont/)
which has 8x16 halfwidth and 16x16 fullwidth bitmap characters.
Other fonts will probably not fit that grid and leave drawing artefacts.
There are no other options for gridded display yet.

I added a 'continuous' mode to look at other fonts. Mostly for
the cases where the gridded display just does not work.
In this mode glyphs will be added to a line while they fit.
Currently you can only move to the next page, there is
no moving back.

Some interesting fonts to look at:
  (Links checked 2013/09/24)
  The above mentioned unifont: http://czyborra.com/unifont/
  Aegean, Akkadian from http://users.teilar.gr/~g1951d/
  Most of the glyphs are outside of the BMP so use
    unicodeview -p 0x100 -m 3,10 'aegean.ttf'
    resp.
    unicodeview -p 0x120 -m 3,10 'akkadian.ttf'
  to view the interesting pages.

  Lots of fonts to find:
    http://www.alanwood.net/unicode/index.html


Usage
-----

   unicodeview [-g] [-c] [-p 'page'] [-m 'charmap'] 'ttf-font-file'

This will open the TTF file and display the glyphs of the font.
'-g' selects gridded view which shows the glyphs
in pages of 256 glyphs each (in a 16x16 grid).
At the top the page number is displayed as 3 hex digits.
The starting page can be given with the '-p' parameter.
Be careful to give hex numbers with the leading '0x'.
Otherwise it will be interpreted as decimal.
Gridded view is the default so '-g' normally is not
necessary.

'-c' selects continuous view mode
which puts glyphs in lines as long as they fit
and continues with the next glyph in new lines and pages.
The display ends at the last glyph and wraps
to the first glyph on the next page. You cannot
move backward in this mode. Use this mode if the
16x16 grid does not give satisfying results.

The '-m' parameter is a pair of numbers 'p,e',
platform-id 'p' and encoding-id 'e' to select a character mapping
in the font. Font files can contain more than one mapping of indices
to glyphs. The default mapping chosen by SDL_ttf
might not be what you want. Most of the time it chooses
'3,1' which is good for 16-bit Unicode (BMP).
Fonts for ancient scripts with code points behond 0x10000
may use '3,10'. So you should try this if your font
contains such glyphs. To find out about the character
mappings contained in a font you can use the utility
'ttfdump'. Look out for 'cmap' in the output.
If you follow this you can then find the corresponding
'SubTable's and see which glyphs are accessible with which
mapping. A resource to find out about the meaning of
the platform and encoding IDs is the header 'ttnameid.h'
of the freetype library.


Key bindings
------------

You can browse to other pages using '<' and '>'
(only '>' works in continuous mode).
The pages wrap at index '0xFFF' in gridded mode
and at the last glyph in continuous mode.
Use 'q' to quit the program.


Build and install
-----------------

To build and install the program use the usual
autoconf commands:

  > ./configure
  > make
  > make install

Building requisites are installed development files of SDL,
of the freetype2 library and of SDL_draw, a small drawing library for SDL.
It also uses a modified version of SDL_ttf 2.0.11, the ttf library
for SDL by Sam Lantinga. The modified sources are included here.
You can find the Copyright for SDL_ttf in the file COPYING_SDL_ttf.

I tested using the following versions of the packages:
  freetype2    -- 2.4.9  (debian 2.4.9-1.1)
  libsdl1.2    -- 1.2.15 (debian 1.2.15-5)
  SDL_draw     -- 1.2.13


Have fun,
Matthias Kievernagel


NEWS
 -*- mode: text; coding: utf-8 -*-
2013-06-26  Matthias Kievernagel  mkiever/at/tlink/dot/de


sdl unicode font view news
--------------------------

2013-09-23: (version 0.03)
    Implemented a 'continuous' mode for glyph display
    to accomodate fonts with large differences in
    character width. This starts display with the first glyph
    of the font and fits as much glyphs in a line
    and as much lines in the window as possible.
    You can only advance (using '>') to the following glyphs
    until arriving at the last one. Next '>' will restart
    with displaying the first glyphs.

2013-07-05:
    With the inclusion of a modified version of SDL_ttf
    unicodeview can now display glyphs outside of the BMP.
    Command line parameters allow to go to a specific page
    directly and to select another character mapping
    (necessary for glyphs outside the BMP).

First packaging 26-06-2013.


SDL_ttf.h
/*
  Modified 2013 by Matthias Kievernagel (mkiever/at/tlink/dot/de)

  SDL_ttf:  A companion library to SDL for working with TrueType (tm) fonts
  Copyright (C) 2001-2012 Sam Lantinga <slouken@libsdl.org>

  This software is provided 'as-is', without any express or implied
  warranty.  In no event will the authors be held liable for any damages
  arising from the use of this software.

  Permission is granted to anyone to use this software for any purpose,
  including commercial applications, and to alter it and redistribute it
  freely, subject to the following restrictions:

  1. The origin of this software must not be misrepresented; you must not
     claim that you wrote the original software. If you use this software
     in a product, an acknowledgment in the product documentation would be
     appreciated but is not required.
  2. Altered source versions must be plainly marked as such, and must not be
     misrepresented as being the original software.
  3. This notice may not be removed or altered from any source distribution.
*/

/* This library is a wrapper around the excellent FreeType 2.0 library,
   available at:
	http://www.freetype.org/
*/

#ifndef _SDL_TTF_H
#define _SDL_TTF_H

#include <ft2build.h>
#include FT_FREETYPE_H

#include <SDL/SDL.h>
#include <SDL/begin_code.h>

/* Set up for C function definitions, even when using C++ */
#ifdef __cplusplus
extern "C" {
#endif

/* Printable format: "%d.%d.%d", MAJOR, MINOR, PATCHLEVEL
*/
#define SDL_TTF_MAJOR_VERSION	2
#define SDL_TTF_MINOR_VERSION	0
#define SDL_TTF_PATCHLEVEL	11

/* This macro can be used to fill a version structure with the compile-time
 * version of the SDL_ttf library.
 */
#define SDL_TTF_VERSION(X)						\
{									\
	(X)->major = SDL_TTF_MAJOR_VERSION;				\
	(X)->minor = SDL_TTF_MINOR_VERSION;				\
	(X)->patch = SDL_TTF_PATCHLEVEL;				\
}

/* Backwards compatibility */
#define TTF_MAJOR_VERSION	SDL_TTF_MAJOR_VERSION
#define TTF_MINOR_VERSION	SDL_TTF_MINOR_VERSION
#define TTF_PATCHLEVEL		SDL_TTF_PATCHLEVEL
#define TTF_VERSION(X)		SDL_TTF_VERSION(X)

/* This function gets the version of the dynamically linked SDL_ttf library.
   it should NOT be used to fill a version structure, instead you should
   use the SDL_TTF_VERSION() macro.
 */
extern DECLSPEC const SDL_version * SDLCALL TTF_Linked_Version(void);

/* ZERO WIDTH NO-BREAKSPACE (Unicode byte order mark) */
#define UNICODE_BOM_NATIVE	0xFEFF
#define UNICODE_BOM_SWAPPED	0xFFFE

/* This function tells the library whether UNICODE text is generally
   byteswapped.  A UNICODE BOM character in a string will override
   this setting for the remainder of that string.
*/
extern DECLSPEC void SDLCALL TTF_ByteSwappedUNICODE(int swapped);

/* The internal structure containing font information */
typedef struct _TTF_Font TTF_Font;

/* Initialize the TTF engine - returns 0 if successful, -1 on error */
extern DECLSPEC int SDLCALL TTF_Init(void);

/* Open a font file and create a font of the specified point size.
 * Some .fon fonts will have several sizes embedded in the file, so the
 * point size becomes the index of choosing which size.  If the value
 * is too high, the last indexed size will be the default. */
extern DECLSPEC TTF_Font * SDLCALL TTF_OpenFont(const char *file, int ptsize);
extern DECLSPEC TTF_Font * SDLCALL TTF_OpenFontIndex(const char *file, int ptsize, long index);
extern DECLSPEC TTF_Font * SDLCALL TTF_OpenFontRW(SDL_RWops *src, int freesrc, int ptsize);
extern DECLSPEC TTF_Font * SDLCALL TTF_OpenFontIndexRW(SDL_RWops *src, int freesrc, int ptsize, long index);

/*xxx*/
/* select char map (if available) */
extern DECLSPEC int SDLCALL TTF_SelectCharMap(TTF_Font *font, int platform_id, int encoding_id);

/* Set and retrieve the font style */
#define TTF_STYLE_NORMAL	0x00
#define TTF_STYLE_BOLD		0x01
#define TTF_STYLE_ITALIC	0x02
#define TTF_STYLE_UNDERLINE	0x04
#define TTF_STYLE_STRIKETHROUGH	0x08
extern DECLSPEC int SDLCALL TTF_GetFontStyle(const TTF_Font *font);
extern DECLSPEC void SDLCALL TTF_SetFontStyle(TTF_Font *font, int style);
extern DECLSPEC int SDLCALL TTF_GetFontOutline(const TTF_Font *font);
extern DECLSPEC void SDLCALL TTF_SetFontOutline(TTF_Font *font, int outline);

/* Set and retrieve FreeType hinter settings */
#define TTF_HINTING_NORMAL    0
#define TTF_HINTING_LIGHT     1
#define TTF_HINTING_MONO      2
#define TTF_HINTING_NONE      3
extern DECLSPEC int SDLCALL TTF_GetFontHinting(const TTF_Font *font);
extern DECLSPEC void SDLCALL TTF_SetFontHinting(TTF_Font *font, int hinting);

/* Get the total height of the font - usually equal to point size */
extern DECLSPEC int SDLCALL TTF_FontHeight(const TTF_Font *font);

/* Get the offset from the baseline to the top of the font
   This is a positive value, relative to the baseline.
 */
extern DECLSPEC int SDLCALL TTF_FontAscent(const TTF_Font *font);

/* Get the offset from the baseline to the bottom of the font
   This is a negative value, relative to the baseline.
 */
extern DECLSPEC int SDLCALL TTF_FontDescent(const TTF_Font *font);

/* Get the recommended spacing between lines of text for this font */
extern DECLSPEC int SDLCALL TTF_FontLineSkip(const TTF_Font *font);

/* Get/Set whether or not kerning is allowed for this font */
extern DECLSPEC int SDLCALL TTF_GetFontKerning(const TTF_Font *font);
extern DECLSPEC void SDLCALL TTF_SetFontKerning(TTF_Font *font, int allowed);

/* Get the number of faces of the font */
extern DECLSPEC long SDLCALL TTF_FontFaces(const TTF_Font *font);

/* Get the font face attributes, if any */
extern DECLSPEC int SDLCALL TTF_FontFaceIsFixedWidth(const TTF_Font *font);
extern DECLSPEC char * SDLCALL TTF_FontFaceFamilyName(const TTF_Font *font);
extern DECLSPEC char * SDLCALL TTF_FontFaceStyleName(const TTF_Font *font);

/* Check wether a glyph is provided by the font or not */
extern DECLSPEC int SDLCALL TTF_GlyphIsProvided(const TTF_Font *font, Uint16 ch);

/* Get the metrics (dimensions) of a glyph
   To understand what these metrics mean, here is a useful link:
    http://freetype.sourceforge.net/freetype2/docs/tutorial/step2.html
 */
/*xxx*/
extern DECLSPEC int SDLCALL TTF_GlyphMetrics(TTF_Font *font, Uint32 ch,
				     int *minx, int *maxx,
                                     int *miny, int *maxy, int *advance);

/* Get the dimensions of a rendered string of text */
extern DECLSPEC int SDLCALL TTF_SizeText(TTF_Font *font, const char *text, int *w, int *h);
extern DECLSPEC int SDLCALL TTF_SizeUTF8(TTF_Font *font, const char *text, int *w, int *h);
extern DECLSPEC int SDLCALL TTF_SizeUNICODE(TTF_Font *font, const Uint16 *text, int *w, int *h);

/* Create an 8-bit palettized surface and render the given text at
   fast quality with the given font and color.  The 0 pixel is the
   colorkey, giving a transparent background, and the 1 pixel is set
   to the text color.
   This function returns the new surface, or NULL if there was an error.
*/
extern DECLSPEC SDL_Surface * SDLCALL TTF_RenderText_Solid(TTF_Font *font,
				const char *text, SDL_Color fg);
extern DECLSPEC SDL_Surface * SDLCALL TTF_RenderUTF8_Solid(TTF_Font *font,
				const char *text, SDL_Color fg);
extern DECLSPEC SDL_Surface * SDLCALL TTF_RenderUNICODE_Solid(TTF_Font *font,
				const Uint16 *text, SDL_Color fg);

/* Create an 8-bit palettized surface and render the given glyph at
   fast quality with the given font and color.  The 0 pixel is the
   colorkey, giving a transparent background, and the 1 pixel is set
   to the text color.  The glyph is rendered without any padding or
   centering in the X direction, and aligned normally in the Y direction.
   This function returns the new surface, or NULL if there was an error.
*/
/*xxx*/
extern DECLSPEC SDL_Surface * SDLCALL TTF_RenderGlyph_Solid(TTF_Font *font,
					Uint32 ch, SDL_Color fg);

/* Create an 8-bit palettized surface and render the given text at
   high quality with the given font and colors.  The 0 pixel is background,
   while other pixels have varying degrees of the foreground color.
   This function returns the new surface, or NULL if there was an error.
*/
extern DECLSPEC SDL_Surface * SDLCALL TTF_RenderText_Shaded(TTF_Font *font,
				const char *text, SDL_Color fg, SDL_Color bg);
extern DECLSPEC SDL_Surface * SDLCALL TTF_RenderUTF8_Shaded(TTF_Font *font,
				const char *text, SDL_Color fg, SDL_Color bg);
extern DECLSPEC SDL_Surface * SDLCALL TTF_RenderUNICODE_Shaded(TTF_Font *font,
				const Uint16 *text, SDL_Color fg, SDL_Color bg);

/* Create an 8-bit palettized surface and render the given glyph at
   high quality with the given font and colors.  The 0 pixel is background,
   while other pixels have varying degrees of the foreground color.
   The glyph is rendered without any padding or centering in the X
   direction, and aligned normally in the Y direction.
   This function returns the new surface, or NULL if there was an error.
*/
extern DECLSPEC SDL_Surface * SDLCALL TTF_RenderGlyph_Shaded(TTF_Font *font,
				Uint16 ch, SDL_Color fg, SDL_Color bg);

/* Create a 32-bit ARGB surface and render the given text at high quality,
   using alpha blending to dither the font with the given color.
   This function returns the new surface, or NULL if there was an error.
*/
extern DECLSPEC SDL_Surface * SDLCALL TTF_RenderText_Blended(TTF_Font *font,
				const char *text, SDL_Color fg);
extern DECLSPEC SDL_Surface * SDLCALL TTF_RenderUTF8_Blended(TTF_Font *font,
				const char *text, SDL_Color fg);
extern DECLSPEC SDL_Surface * SDLCALL TTF_RenderUNICODE_Blended(TTF_Font *font,
				const Uint16 *text, SDL_Color fg);

/* Create a 32-bit ARGB surface and render the given glyph at high quality,
   using alpha blending to dither the font with the given color.
   The glyph is rendered without any padding or centering in the X
   direction, and aligned normally in the Y direction.
   This function returns the new surface, or NULL if there was an error.
*/
extern DECLSPEC SDL_Surface * SDLCALL TTF_RenderGlyph_Blended(TTF_Font *font,
						Uint16 ch, SDL_Color fg);

/* For compatibility with previous versions, here are the old functions */
#define TTF_RenderText(font, text, fg, bg)	\
	TTF_RenderText_Shaded(font, text, fg, bg)
#define TTF_RenderUTF8(font, text, fg, bg)	\
	TTF_RenderUTF8_Shaded(font, text, fg, bg)
#define TTF_RenderUNICODE(font, text, fg, bg)	\
	TTF_RenderUNICODE_Shaded(font, text, fg, bg)

/* Close an opened font file */
extern DECLSPEC void SDLCALL TTF_CloseFont(TTF_Font *font);

/* De-initialize the TTF engine */
extern DECLSPEC void SDLCALL TTF_Quit(void);

/* Check if the TTF engine is initialized */
extern DECLSPEC int SDLCALL TTF_WasInit(void);

/* Get the kerning size of two glyphs */
extern DECLSPEC int TTF_GetFontKerningSize(TTF_Font *font, int prev_index, int index);

/* We'll use SDL for reporting errors */
#define TTF_SetError	SDL_SetError
#define TTF_GetError	SDL_GetError

/*xxx*/
/* get FreeType Face from TTF_Font */
extern DECLSPEC FT_Face TTF_GetFreeTypeFace(TTF_Font *font);

/* Ends C function definitions when using C++ */
#ifdef __cplusplus
}
#endif
#include <SDL/close_code.h>

#endif /* _SDL_TTF_H */


unipageview.h
#ifndef INC_UNIPAGEVIEW_H
#define INC_UNIPAGEVIEW_H

/* ------------------------------------------------------------------------- */
/* unipageview.h - this file is part of unicodeview                          */
/* Copyright 2012, 2013                                                      */
/* Matthias Kievernagel                                                      */
/*                                                                           */
/* This program is free software; you can redistribute it and/or modify      */
/* it under the terms of the GNU General Public License version 2 as         */
/* published by the Free Software Foundation.                                */
/*                                                                           */
/* This program is distributed in the hope that it will be useful,           */
/* but WITHOUT ANY WARRANTY; without even the implied warranty of            */
/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the             */
/* GNU General Public License for more details.                              */
/*                                                                           */
/* You should have received a copy of the GNU General Public License         */
/* along with this program.  If not, see <http://www.gnu.org/licenses/>.     */
/* ------------------------------------------------------------------------- */

/* unicode page view */
/* shows 256 glyphs of a font in a 8x8 matrix */


#include <SDL/SDL.h>
#include "SDL_ttf.h"


struct unipageview {
  SDL_Surface * parent;
  SDL_Surface * view;
  SDL_Rect srect;
  int gridded;
  int dist;
  unsigned int page_base;
  unsigned int first_cpt;
  unsigned int first_index;
  unsigned int last_cpt;
  unsigned int last_index;
  TTF_Font * font;
};


extern struct unipageview * upv_create(
  SDL_Surface * parent, SDL_Rect * r);
extern int upv_init(struct unipageview * upv);
extern void upv_render(struct unipageview * upv);
extern void upv_setfont(struct unipageview * upv, TTF_Font * font);
extern unsigned int upv_getbase(struct unipageview * upv);
extern void upv_setbase(struct unipageview * upv, unsigned int base);
extern void upv_setgridded(struct unipageview * upv, int gridded);
extern int upv_gridded(struct unipageview * upv);
extern void upv_next(struct unipageview * upv);
extern void upv_setdistance(struct unipageview * upv, int dist);

#endif


unicodeview.c

/* ------------------------------------------------------------------------- */
/* unicodeview.c - this file is part of unicodeview                          */
/* Copyright 2012, 2013                                                      */
/* Matthias Kievernagel                                                      */
/*                                                                           */
/* This program is free software; you can redistribute it and/or modify      */
/* it under the terms of the GNU General Public License version 2 as         */
/* published by the Free Software Foundation.                                */
/*                                                                           */
/* This program is distributed in the hope that it will be useful,           */
/* but WITHOUT ANY WARRANTY; without even the implied warranty of            */
/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the             */
/* GNU General Public License for more details.                              */
/*                                                                           */
/* You should have received a copy of the GNU General Public License         */
/* along with this program.  If not, see <http://www.gnu.org/licenses/>.     */
/* ------------------------------------------------------------------------- */

/* main program for viewing unicode fonts */


#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <SDL/SDL.h>
#include "SDL_ttf.h"

#include <ft2build.h>
#include FT_FREETYPE_H

#include "unipageview.h"


SDL_Surface * _pm_screen = NULL;

/* unicode window */
struct unipageview * _upview = NULL;

/* The font that's going to be used */
TTF_Font * _pm_font = NULL;


static void _usage() {
    fprintf(stderr,
        "Usage: unicodeview "
        " [-g] [-c] [-p 'page'] [-m 'charmap'] 'fontfilename'\n");
    fprintf(stderr, " -g: display in gridded mode (default)\n");
    fprintf(stderr, " -c: display in continuous mode\n");
    fprintf(stderr,
        " -p 'page': Start display on page 'page' in gridded mode.\n");
    fprintf(stderr,
        "            'page' must be given in hex format, i.e. '0xnnn'\n");
    fprintf(stderr,
        " -m 'charmap': Select 'charmap' in the font for indexing.\n");
    fprintf(stderr,
        "               'charmap' has the format 'number,number'.\n");
    fprintf(stderr,
        "               See README for more details.\n");
}


static char * _init_sdl(
        char * fontfilename, int fontsize, int platform_id, int encoding_id) {
    int res;
    FT_Face face;
    FT_UInt index;
    FT_ULong charcode;

    /* Start SDL */
    res = SDL_Init(SDL_INIT_VIDEO);
    if ( res == -1 ) {
        return SDL_GetError();
    }

    /* setup screen */
    _pm_screen = SDL_SetVideoMode(360, 390, 32, SDL_SWSURFACE);
    if ( _pm_screen == NULL ) {
        return SDL_GetError();
    }

    /* Initialize SDL_ttf */
    if ( TTF_Init() == -1 ) {
        return TTF_GetError();
    }

    /* Open the font */
    /* _pm_font = TTF_OpenFont ( "DejaVuSansMono.ttf", 16 ); */
    /* _pm_font = TTF_OpenFont ( "LinBiolinum_Rah.ttf", 16 ); */
    /* _pm_font = TTF_OpenFont("unifont.ttf", 16); */
    _pm_font = TTF_OpenFont(fontfilename, fontsize);
    if ( _pm_font == NULL ) {
        return TTF_GetError();
    }

    /* set specified platform/encoding charmap, otherwise use default */
    if ( platform_id != -1 && encoding_id != -1 ) {
        res = TTF_SelectCharMap(_pm_font, platform_id, encoding_id);
        /* printf("TTF_SelectCharMap %d\n", res); */
    }

    /* debug output:
    printf("faces %ld\n", TTF_FontFaces(_pm_font));
    printf("fixed %d\n", TTF_FontFaceIsFixedWidth(_pm_font));
    printf("kerning %d\n", TTF_GetFontKerning(_pm_font));
    printf("height %d\n", TTF_FontHeight(_pm_font));
    printf("ascent %d\n", TTF_FontAscent(_pm_font));
    printf("descent %d\n", TTF_FontDescent(_pm_font));
    printf("lineskip %d\n", TTF_FontLineSkip(_pm_font));
    */

    /* test iterating char codes:
    face = TTF_GetFreeTypeFace(_pm_font);
    charcode = FT_Get_First_Char(face, &index);
    while ( index != 0 ) {
        printf("char %08lx/%d\n", charcode, index);
        charcode = FT_Get_Next_Char(face, charcode, &index);
    }
    */

    SDL_EnableUNICODE(1);

    return NULL;
}

static void _teardown_sdl () {
    /* Close the font that was used */
    TTF_CloseFont(_pm_font);
    
    /* Quit SDL_ttf + SDL */
    TTF_Quit();
    SDL_Quit();
}

static void _pm_mainloop () {
    SDL_Event event;
    int _quit = 0;
    int base;
    int cpt;
    int dirty = 1;
    int minx,miny,maxx,maxy,adv;

    while ( _quit != 1 ) {
        while ( SDL_PollEvent ( &event )) {
            switch ( event.type ) {
            case SDL_KEYDOWN:
                /*
                printf("key down unicode (%hd)\n", event.key.keysym.unicode);
                */
                if ( event.key.keysym.unicode == (int)'q' ) {
                    _quit = 1;
                }
                /* debug output glyph metrics
                else if ( event.key.keysym.unicode == (int)'p' ) {
                    base = upv_getbase(_upview);
                    cpt = (base << 8) + 32;
                    TTF_GlyphMetrics(
                        _pm_font, cpt, &minx, &maxx, &miny, &maxy, &adv);
                    printf("[%04x]x(%d/%d)y(%d/%d)a(%d)\n", cpt,
                        minx, maxx, miny, maxy, adv);
                } */
                else if ( event.key.keysym.unicode == (int)'<' ) {
                    base = upv_getbase(_upview);
                    upv_setbase(_upview, base - 1);
                    upv_render(_upview);
                    dirty = 1;
                }
                else if ( event.key.keysym.unicode == (int)'>' ) {
                    if ( upv_gridded(_upview)) {
                        base = upv_getbase(_upview);
                        upv_setbase(_upview, base + 1);
                    } else {
                        upv_next(_upview);
                    }

                    upv_render(_upview);
                    dirty = 1;
                }
                break;

            case SDL_QUIT:
                _quit = 1;
                break;

            default:
                break;
            }
        }

        if ( dirty == 1 ) {
            SDL_Flip ( _pm_screen );
            dirty = 0;
        }

        SDL_Delay ( 50 );
    }
}

int main ( int argc, char * argv[] ) {
    char * err;
    int platform_id = -1;
    int encoding_id = -1;
    int res;
    int page = 0;
    int opt;
    char * fontname = NULL;
    int gridded = 1;

    while (( opt = getopt ( argc, argv, "-gcp:m:" )) != -1 ) {
        switch ( opt ) {
        case 'g':
            gridded = 1;
            break;
        case 'c':
            gridded = 0;
            break;
        case 'p':
            res = sscanf(optarg, "%i", &page);
            if ( res != 1 || page < 0 || page > 0xFFF ) {
                fprintf(stderr, "Illegal page (%s)\n", optarg);
                _usage();
                exit(1);
            }
            break;
        case 'm':
            res = sscanf(optarg, "%d , %d", &platform_id, &encoding_id);
            if ( res != 2 || platform_id < 0 || encoding_id < 0 ) {
                fprintf(stderr,
                    "Illegal charmap specification (%s)\n", optarg);
                _usage();
                exit(1);
            }
            break;
        case 1:
            fontname = optarg;
            break;
        default:
            /* printf ( "opt (%c,%d)\n", opt, opt ); */
            _usage();
            exit(1);
            break;
        }
    }

    if ( fontname == NULL ) {
        printf ( "font missing\n" );
        _usage();
        exit(1);
    }

    err = _init_sdl(fontname, 64, platform_id, encoding_id);
    if ( err != NULL ) {
        fprintf(stderr, "Init failed: %s\n", err);
        exit(1);
    }

    SDL_WM_SetCaption("Unicode font viewer", "Unicode font viewer");

    SDL_Rect offset;
    offset.x = 10;
    offset.y = 10;
    offset.w = 340;
    offset.h = 370;
    _upview = upv_create(_pm_screen, &offset);
    upv_setfont(_upview, _pm_font);
    upv_setgridded(_upview, gridded);
    upv_init(_upview);
    /* in gridded mode: set page */
    if ( gridded ) {
        upv_setbase(_upview, page);
    }
    upv_render(_upview);

    _pm_mainloop ();
    
    _teardown_sdl ();
    return 0;
}
